src/lib/list/selection-list.ts

```diff
@@ -0,0 +1,228 @@
+
+  constructor(private _element: ElementRef) { }
+
+  ngAfterContentInit(): void {
+    this._keyManager = new FocusKeyManager(this.options).withWrap();
+
+    // Prevents the selection-list from capturing focus and redirecting
+    // it back to the first option when the user tabs out.
+    this._tabOutSubscription = this._keyManager.tabOut.subscribe(() => {
+      this._tabIndex = -1;
+      setTimeout(() => this._tabIndex = 0);
+    });
+
+    // Go ahead and subscribe all of the initial options
+    this._subscribeOptions(this.options);
+    // When the list changes, re-subscribe
+    this._optionSubscription =
+      this.options.changes.subscribe((options: QueryList<MdListOption>) => {
+        this._subscribeOptions(options);
+      });
+  }
+
+  ngOnDestroy(): void {
+    if (this._tabOutSubscription) {
+      this._tabOutSubscription.unsubscribe();
+    }
+
+    if (this._optionSubscription) {
+      this._optionSubscription.unsubscribe();
+    }
+  }
+
+  /**
+   * Whether or not this option is selectable. When a option is not selectable,
+   * it's selected state is always ignored.
+   */
+  @Input()
+  get selectable(): boolean { return this._selectable; }
+  set selectable(value: boolean) {
+    this._selectable = coerceBooleanProperty(value);
+  }
+
+  focus() {
+    this._element.nativeElement.focus();
+  }
+
+  /** Passes relevant key presses to our key manager. */
+  keydown(event: KeyboardEvent) {
+    switch (event.keyCode) {
+      case SPACE:
+        this._toggleSelectOnFocusedOption();
+        // Always prevent space from scrolling the page since the list has focus
+        event.preventDefault();
+        break;
+      default:
+        this._keyManager.onKeydown(event);
+    }
+  }
+
+  /** Toggles the selected state of the currently focused option. */
+  protected _toggleSelectOnFocusedOption(): void {
+    if (!this.selectable) {
+      return;
+    }
+
+    let focusedIndex = this._keyManager.activeItemIndex;
+
+    if (focusedIndex != null && this._isValidIndex(focusedIndex)) {
+      let focusedOption: MdListOption = this.options.toArray()[focusedIndex];
+
+      if (focusedOption) {
+        focusedOption.toggle();
+      }
+    }
+  }
+
+
+  /**
+   * Iterate through the list of options and add them to our list of
+   * subscribed options.
+   *
+   * @param options The list of options to be subscribed.
+   */
+  protected _subscribeOptions(options: QueryList<MdListOption>): void {
+    options.forEach(option => this._addOption(option));
+  }
+
+  /**
+   * Add a specific option to our subscribed list. If the option has
+   * already been subscribed, this ensures it is only subscribed
+   * once.
+   *
+   * @param option The option to be subscribed (or checked for existing
+   * subscription).
+   */
+  protected _addOption(option: MdListOption) {
+    // If we've already been subscribed to a parent, do nothing
+    if (this._subscribed.has(option)) {
```

Comment:
Instead of keeping a list of options that were previously subscribed, it might be easier to merge all the option subscriptions into one stream. Then you can just switchMap to a new merged stream every time the option list changes. Check out how we do this in the autocomplete [here](https://github.com/angular/material2/blob/master/src/lib/autocomplete/autocomplete-trigger.ts#L234) and [here](https://github.com/angular/material2/blob/master/src/lib/autocomplete/autocomplete-trigger.ts#L384).